File: sliding_window.py
import random
from datetime import datetime

def find_max_sum_of_K_brute_force(arr, K):
    max_sum = float('-inf')
    for i in range(len(arr)-K+1):
        current_sum = sum(arr[i:i+K])            
        max_sum = max(current_sum, max_sum)            
    return max_sum

def find_max_sum_of_K_sliding_window(arr, K):
    current_sum = sum(arr[0:K])
    
    max_sum = current_sum
    for i in range(1, len(arr)-K+1):
        current_sum -= arr[i-1]
        current_sum += arr[i-1 + K]
        max_sum = max(current_sum, max_sum)            
    return max_sum
        

arr = random.sample(range(0, 100000), 99999)
K = 999
start = datetime.now()
max_br = find_max_sum_of_K_brute_force(arr, K)
print("max_br: {} ({})".format(max_br, (datetime.now()-start).total_seconds()))

start = datetime.now()
max_sw = find_max_sum_of_K_sliding_window(arr, K)
print("max_sw: {} ({})".format(max_sw, (datetime.now()-start).total_seconds()))

File: sliding_window.cpp
#include <vector>
#include <iostream>
#include <algorithm>
#include <numeric>
#include <limits>
#include <cmath>
#include "measure.hpp"

static int64_t find_max_sum_of_K_brute_force(std::vector<int>& arr, const size_t K)
{
    int64_t max_sum = std::numeric_limits<int64_t>::infinity();
    for(size_t i=0; i<arr.size()-K; i++)
    {
        int64_t sum = std::accumulate(begin(arr)+i, begin(arr)+i+K, 0);
        max_sum = std::max(sum, max_sum);            
    }

    return max_sum;
}

static uint64_t find_max_sum_of_K_sliding_window(std::vector<int>& arr, const size_t K)
{
    int64_t sum = std::accumulate(begin(arr), begin(arr)+K, 0);
    auto max_sum = sum;
    for(int i=1; i<arr.size()-K+1; i++)
    {
        sum -= arr[i-1];
        sum += arr[i+K-1];        
        max_sum = std::max(sum, max_sum);
    }

    return max_sum;
}

int main(void)
{
    std::vector<int> v(999999);
    constexpr size_t K = 999;
    srand(12314);
    for(auto& e : v)
    {
        e = random() % 10000;        
    }

    perform_print_duration("max_br: ", find_max_sum_of_K_brute_force, v, K);
    perform_print_duration("max_sw: ", find_max_sum_of_K_sliding_window, v, K);

    return 0;
}